### DISCUSSION TITLE
formating variables with dots and slash in name

### DISCUSSION THREAD

**Discussion by prune998** [TomWright/dasel#228] [Q&A] [answered]
Hi all,

I'm trying to use Dasel to select (filter) one `annotation` out of a `kubectl get` commands. The annotation is not always present in all the objects, so I thought I would solve the issue by using a `--format` option, but so far I failed.

`Kubectl` command is:

```
kubectl get ingress -A -o json
```

which returns a json with:

```
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "networking.k8s.io/v1",
            "kind": "Ingress",
            "metadata": {
                "annotations": {
                    "example": "value",
                    "kubectl.kubernetes.io/last-applied-configuration": "{\"escaped\":\"jsonValues\"}"
...
```

I first tried to only use a select:
```
k get ing -A -o json | dasel -p json select -m  -s 'items.[*].metadata.annotations.kubectl\.kubernetes\.io/last-applied-configuration'

Error: could not query multiple node: could not find multiple value: no value found for selector: .kubectl.kubernetes.io/last-applied-configuration
```

Using `--null` only prints `null`.

Using format I tried different variations of:

```
k get ing -A -o json | dasel -p json select -m  -s 'items.[*].metadata.annotations' --format '{{ if .kubectl\.kubernetes\.io\/last-applied-configuration }}{{ select ".kubectl.kubernetes.io/last-applied-configuration" }}{{ else }}-{{ end}}'
```

But they all fail on the `if` because it contains `.` or `\` or `"` ...
Any way of achieving this ? 

Thanks

**Comment by TomWright** [comment, marked as answer]
You've got the right idea, but with the golang templating you need to use the `index` function to access map keys: https://pkg.go.dev/text/template#hdr-Functions
> index
>	Returns the result of indexing its first argument by the
>	following arguments. Thus "index x 1 2 3" is, in Go syntax,
>	x[1][2][3]. Each indexed item must be a map, slice, or array.

For context, the format string is a plain golang `text/template` and is processed outside of the normal dasel selector syntax. You could probably use the `select` func in the condition to use dasel syntax but in the meantime here's what I came up with:
```bash
$ echo '{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "networking.k8s.io/v1",
            "kind": "Ingress",
            "metadata": {
                "annotations": {
                    "example": "value",
                    "kubectl.kubernetes.io/last-applied-configuration": "{\"escaped\":\"jsonValues\"}"
                }
            }
        },
        {
            "apiVersion": "networking.k8s.io/v1",
            "kind": "Ingress",
            "metadata": {
                "annotations": {
                    "example": "value"
                }
            }
        }
    ]
}' | dasel -p json -m -s 'items.[*].metadata.annotations' --format '{{ if index . "kubectl.kubernetes.io/last-applied-configuration" }}{{ toJson . }}{{ else }}-{{ end}}'
{"example":"value","kubectl.kubernetes.io/last-applied-configuration":"{\"escaped\":\"jsonValues\"}"}
-
```

https://dasel.tomwright.me/s/3b29e743-00f1-48f3-b3e6-884289502168

If you want pretty JSON there is a `toPrettyJson` func you could use instead.

I can see the value in checking the presence of keys so perhaps that will come in the future.
